### ISSUE TITLE
Feature parity with current notebook

### ISSUE THREAD

**Issue by jasongrout** [jupyter/jupyter-js-notebook#87]
This is a meta issue listing high-level features of the current notebook we need to address in a new implementation. We may decide that a feature is not worth implementing, or that we want to implement it differently.
- [x] Widgets (#34) (changes how widgets are displayed)
- [x] Multiple-cell selection
- [x] keyboard navigation / shortcuts
  - [x] continuous keyboard traversal of cells with up/down in edit mode, rather than stopping hard at cell boundaries, forcing the user to 'pop out' to command mode to move between cells. Basically, when you press the down arrow at the bottom of a cell, it moves to the next cell in edit mode if it is a code cell. I think the current notebook also will change a markdown cell to edit mode, which to me is confusing. Similarly for the up arrow when at the top of a cell.
- [ ] printing stylesheet
- [ ] convert to variety of formats (current notebook "File > Download as" menu)
- [ ] Find and replace
- [ ] a variety of commands (compare with command list in current notebook)
- [ ] help links
- [ ] `?` and `??` introspection (we have `?`, but seeing `??` requires opening the inspector manually -  perhaps `??` should expose a link to open the inspector to guide the user?)
- [ ] tooltip help (function signatures, help, etc.) https://github.com/jupyterlab/jupyterlab/issues/714
- [ ] edit notebook and cell metadata
- [x] Changing kernels
- [ ] Dealing with large output (in classic notebook, clicking on output prompt switches output between hidden, constrained-height, and full-height)
- [x] Notebook trust features
- [x] tab completion in editor
- [ ] Cell toolbars https://github.com/jupyterlab/jupyterlab/issues/902
- [ ] undeletable cells
- [x] input/raw_input working (for things like pdb)

From @ellisonbg below:
- Pager (this should be a separate area on the page rather than in the notebook widget itself)
- Output scrolling and collapsing need to be reworked, keeping in mind that we also want to allow input to be collapsed. We should have a nice unified UI for doing both of these things.
- Cell tooltips (shift-tab) should be unified with regular tab completion (tab) into a single nice UI/UX.

**Comment by jasongrout** [comment]
CC @ellisonbg, @blink1073, @rgbkrk ... wow, this will be a long list. I'm just going to send an email to everyone.

**Comment by blink1073** [comment]
I would add the output area prompt overlay (which I didn't even know existed until I was trying to clone the notebook).

**Comment by blink1073** [comment]
Plus the ability to edit cell metadata.

**Comment by blink1073** [comment]
And changing kernels (which I'm working on now).

**Comment by jasongrout** [comment]
what is the output area prompt overlay?

**Comment by blink1073** [comment]
I'll make a screencast.

**Comment by jasongrout** [comment]
Is it the clicking on the output area to change the max-height of the output?

**Comment by blink1073** [comment]
![notebook-output-overlay](https://cloud.githubusercontent.com/assets/2096628/13675249/57fa3c66-e6a6-11e5-9a10-ce4da6150e5d.gif)

**Comment by jasongrout** [comment]
yep, that's really useful. In Sage, we also had a mode that would wrap text.

**Comment by jdfreder** [comment]
Tab completion?

**Comment by jasongrout** [comment]
Thanks

**Comment by jhamrick** [comment]
- Ability to edit notebook metadata
- Cell toolbars
- Undeletable cells

**Comment by jasongrout** [comment]
undeletable cells? We have that? How do you do that?

**Comment by Carreau** [comment]
> undeletable cells? We have that? How do you do that?

Not in UI. If you tag a cell with the right metadata it cannot be deleted.

**Comment by Carreau** [comment]
- Tooltips (shift tab)

**Comment by jhamrick** [comment]
> Not in UI. If you tag a cell with the right metadata it cannot be deleted.

Yep, exactly. This is a really important feature for nbgrader to prevent the cell metadata that nbgrader relies upon from being lost.

**Comment by jhamrick** [comment]
Ooh, here's another one -- not sure if this falls under widgets but making sure `input` and `raw_input` work (and relatedly, debuggers like ipdb).

**Comment by jasongrout** [comment]
oh yeah, right!

**Comment by sccolbert** [comment]
Do debuggers work currently?

**Comment by jdfreder** [comment]
ipdb works

**Comment by blink1073** [comment]
Can we agree on a subset of these for an alpha release, to get earlier feedback?

**Comment by rgbkrk** [comment]
![screenshot 2016-03-10 15 13 01](https://cloud.githubusercontent.com/assets/836375/13684688/7fe22b40-e6d2-11e5-8c6d-e9d87ad2b02a.png)

**Comment by minrk** [comment]
@blink1073 for getting it turned on in master, it might be ready now. We don't need anything close to all this for people to start playing with it, if you want to make a PR re-enabling notebooks in Lab.

@jasongrout We should be careful a bit with parity, because we don't necessarily want to copy everything over. Some things like the output-collapsing UI do not need to be inherited as-is, or even mimicked. We can start from more basic requirements like "Dealing with large output" and work from there.

**Comment by blink1073** [comment]
@minrk, I've got a bit more that I'd like to do before making a PR against master, hopefully early next week.

**Comment by jasongrout** [comment]
@minrk - definitely.  My idea in making the list isn't that we necessarily have to implement _exactly_ what the current notebook has, but that these are the things that people are expecting and we should address. However, we definitely can rethink them. For example, we are changing how widgets are displayed. I can see that wasn't clear from the title, so I'll modify the wording there.

**Comment by blink1073** [comment]
Round 1: https://github.com/jupyter/notebook/pull/1201

**Comment by ellisonbg** [comment]
I agree with @minrk that there are some things that we should not implement with the same UI/UX as the existing notebook:
- Pager (this should be a separate area on the page rather than in the notebook widget itself)
- Output scrolling and collapsing need to be reworked, keeping in mind that we also want to allow input to be collapsed. We should have a nice unified UI for doing both of these things.
- Cell tooltips (shift-tab) should be unified with regular tab completion (tab) into a single nice UI/UX.

**Comment by Carreau** [comment]
> Cell tooltips (shift-tab) should be unified with regular tab completion (tab) into a single nice UI/UX.

We should be able to do that in terminal console as well with prompt-toolkit.

**Comment by fperez** [comment]
I think many folks are going to really be thrown in for a loop in their workflows if we don't offer a way to open a notebook by itself in a single tab with a stable URL. It's extremely useful and it lets people continue using their muscle memory (tab shifting in the browser via keyboard) and tab management tools.

I'm finding it very hard to adapt to not having that (even though I like the combined lab layout also for other things, I'm only suggesting we need the _option_ of a single-notebook-at-url).  I suspect others may encounter the same in everyday work...

**Comment by fperez** [comment]
Another one: continuous keyboard traversal of cells with up/down in edit mode, rather than stopping hard at cell boundaries, forcing the user to 'pop out' to command mode to move between cells.  This makes the editing workflow as it currently works ('hard stops') very painful.

**Comment by jasongrout** [comment]
@fperez - would it be enough to have a button to maximize the current notebook, which maximizes the current tab so that it fills the dock panel or the browser window? What are the use cases for having the URL reflect a single document?

_Edit_: We could provide a right-click -> Open in new tab (or maybe middle click on the filename would open a new tab, to keep the browser paradigm?). However, we need to think through how this works with the plugin and extension system. The extensions right now are hooked together in the Lab application, but we'd want the same plugins registered and working in your single-page view too. Perhaps a right-click -> open in new tab could open a new JupyterLab with the single notebook maximized by default...

**Comment by ellisonbg** [comment]
The point @fperez brings up is super important and worth of its own issue:

https://github.com/jupyter/jupyter-js-notebook/issues/158

A short summary is that this feature is absolutely critical.

**Comment by ellisonbg** [comment]
And another issue about keyboard navigation in the dock panel: https://github.com/jupyter/jupyter-js-notebook/issues/159

**Comment by blink1073** [comment]
@jasongrout, do you mind taking a pass through the list above and checking off the things we have hit, and then I'll create separate issues on jupyterlab with Notebook and Feature Parity labels for the ones we have not already tagged there?

**Comment by jasongrout** [comment]
@blink1073 - I checked off a few and added a few references to existing issues.

**Comment by blink1073** [comment]
- What is the printed stylesheet?

- Convert to a variety of formats: https://github.com/jupyterlab/jupyterlab/issues/964
- Find and replace: https://github.com/jupyterlab/jupyterlab/issues/1074
- Commands: https://github.com/jupyterlab/jupyterlab/issues/1306
- Help links: https://github.com/jupyterlab/jupyterlab/issues/1307
- ? and ?? are handled by redirecting pager output
- Metadata are handled by https://github.com/jupyterlab/jupyterlab/issues/1308 and jupyterlab/jupyterlab#902
- Large outputs: https://github.com/jupyterlab/jupyterlab/issues/1309
- Undeletable cells: https://github.com/jupyterlab/jupyterlab/issues/1310

**Comment by jhamrick** [comment]
Related to undeleteable cells is now also the read-only cell metadata; I'm not sure if that was handled here or not yet: https://github.com/jupyter/notebook/issues/1482

**Comment by blink1073** [comment]
Thanks @jhamrick, I opened https://github.com/jupyterlab/jupyterlab/issues/1312 to track that one.

**Comment by jasongrout** [comment]
> What is the printed stylesheet?

I think it's CSS for the print media type - so you can hit the browser's print button and get a nice looking printout. I'm not sure what that would look like for lab, though, given multiple notebooks on a page. Perhaps that can be shelved in the backlog until we have a good way to view a notebook on a single page, or at least until we can tear notebooks out to child windows (and then hit print..)

**Comment by blink1073** [comment]
We could offer up the most recently focused document for printing.

**Comment by blink1073** [comment]
(In theory, I'm not familiar with using the print media type).

**Comment by blink1073** [comment]
Tracked in https://github.com/jupyterlab/jupyterlab/issues/1314, marked as 1.0 so we at least look at it.

**Comment by jasongrout** [comment]
It's just another selector in css - https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Getting_started/Media.

Perhaps we could do something like that. If there is a CSS class applied to the currently active document, we could do something like `@media print .document:not(.active-document) {display: none}` or something to hide everything except the current document. Interesting...
